### PR TITLE
Refactored Codebase

### PR DESCRIPTION
**Pull request by Fruity-Grebbles**
Refactored codebase. Provides the following features:

- Files with humans can now be saved an loaded without losing editing ability
- Codebase is heavily documented and more readable
- API is usable for expansion

Known Issues:

- ~~New humans inherit previous skeletons~~

### CODE REVIEW

File: exts/siborg.create.human/siborg/create/human/extension.py [hunk 1]
```diff
@@ -1,25 +1,211 @@
...
 
 
 class MakeHumanExtension(omni.ext.IExt):
     # ext_id is current extension id. It can be used with extension manager to query additional information, like where
     # this extension is located on filesystem.
 
     def on_startup(self, ext_id):
+
+        # subscribe to stage events
+        # TODO start listening to stage events only when the window is open,
```
[the remark this answers is not in the document]
Comment (Fruity-Grebbles) [reply exts/siborg.create.human/siborg/create/human/extension.py:22]: This is resolved. I'll make an issue to clean up the TODO comments

File: exts/siborg.create.human/siborg/create/human/extension.py [hunk 2]
```diff
@@ -1,25 +1,211 @@
...
+            self._on_selection_changed()
+
+    def _on_selection_changed(self):
+        # get the current selection and stage
+        selection = self._selection.get_selected_prim_paths()
+        stage = self._usd_context.get_stage()
+        print(f"== selection changed with {len(selection)} items")
+
+        if selection and stage:
+            # TODO reject multiple selection
```
[the remark this answers is not in the document]
Comment (Fruity-Grebbles) [reply exts/siborg.create.human/siborg/create/human/extension.py:57]: Multiple selection is now caught and handled

File: exts/siborg.create.human/siborg/create/human/human.py [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+                # as a proxy with the type as the type
+                if type != "proxymeshes" and type != "clothes":
+                    MHCaller.add_proxy(path, type)
+                else:
+                    # Add every proxy in the "clothes" or "proxymeshes" subdictionary
+                    # In this case, name is unused
+                    for name, path in proxies[type].items():
+                        MHCaller.add_proxy(path, type)
+        # TODO Proxy list is not updated in the UI
+        # TODO this is slow, and should be optimized
```
[the remark this answers is not in the document]
Comment (Fruity-Grebbles) [reply exts/siborg.create.human/siborg/create/human/human.py:374]: 'This' refers to UI update. Both macros and proxies update relatively quickly now.

File: exts/siborg.create.human/siborg/create/human/human.py [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+        bindings : list of `UsdSkel.BindingAPI`
+            USD bindings between meshes and skeleton
+        joint_names : list of str
+            Unique, plaintext names of all joints in the skeleton in USD
+            (breadth-first) order.
+        joint_paths : list of str
+            List of the full usd path to each joint corresponding to the skeleton to bind to
+        """
+
+        # TODO Create weights only once per mesh per skeleton, and reuse them
```
[the remark this answers is not in the document]
Comment (Fruity-Grebbles) [reply exts/siborg.create.human/siborg/create/human/human.py:396]: This was leftover from you. I'm not sure how deeply we explored it, but the method of mesh re-import with each modification precludes any stored weights (unless we store them elsewhere in the USD, but that might be going pretty deep for a hack)

**Comment by Fruity-Grebbles** [comment]
Humans no longer inherit the previous skeleton

**Review by cadop — APPROVED** [review]

**Pull request merged**